### PR TITLE
Richfr/ Add windowsAddress support to WslcCreateContainer() API

### PR DESCRIPTION
**Pull request by 1wizkid**
## PR Checklist

- [ ] **Closes:** Link to issue #xxx
- [ ] **Communication:** I've discussed this with core contributors already. If work hasn't been agreed, this work might be rejected
- [ ] **Tests:** Added/updated if needed and all pass
- [ ] **Localization:** All end user facing strings can be localized
- [ ] **Dev docs:** Added/updated if needed
- [ ] **Documentation updated:** If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/wsl/) and link it here: #xxx

### CODE REVIEW

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 1]
```diff
@@ -555,12 +557,32 @@ try
...
+                    THROW_HR_IF_NULL(E_UNEXPECTED, inet_ntop(AF_INET6, &addr6->sin6_addr, addrBuf, sizeof(addrBuf)));
+                    convertedPort.Family = AF_INET6;
+                }
+                bindingAddressStrings[i] = addrBuf;
+                convertedPort.BindingAddress = bindingAddressStrings[i].c_str();
+            }
+            else
+            {
+                convertedPort.Family = AF_INET;
+                convertedPort.BindingAddress = "127.0.0.1";
```
Comment (JohnMcPMS) [inline src/windows/WslcSDK/wslcsdk.cpp]: Should we just leave these unset if not provided?  If the default behavior is "all addresses" this will force localhost only when no address is provided.

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 2]
```diff
@@ -693,7 +715,11 @@ try
 
     for (uint32_t i = 0; i < portMappingCount; ++i)
     {
-        RETURN_HR_IF(E_NOTIMPL, portMappings[i].windowsAddress != nullptr);
+        if (portMappings[i].windowsAddress != nullptr)
+        {
+            const auto family = portMappings[i].windowsAddress->ss_family;
+            RETURN_HR_IF(E_INVALIDARG, family != AF_INET && family != AF_INET6);
+        }
         RETURN_HR_IF(E_NOTIMPL, portMappings[i].protocol != 0);
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp:812]: We currently don't support UDP so OK with code as is. Will update once UDP is supported.

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 1 above)
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Yes. I'll remove the default settings and let the runtime deal with it. User will need to specify local host when they want it.

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 3]
```diff
@@ -544,9 +544,11 @@ try
     }
 
     std::unique_ptr<WSLCPortMapping[]> convertedPorts;
+    std::vector<std::string> bindingAddressStrings;
     if (internalContainerSettings->ports && internalContainerSettings->portsCount)
     {
         convertedPorts = std::make_unique<WSLCPortMapping[]>(internalContainerSettings->portsCount);
+        bindingAddressStrings.resize(internalContainerSettings->portsCount);
         for (uint32_t i = 0; i < internalContainerSettings->portsCount; ++i)
         {
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp:667]: Valid point, but I don't think warrants a change. The callee copies the data so we are good.

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 4]
```diff
@@ -555,12 +557,42 @@ try
             convertedPort.HostPort = internalPort.windowsPort;
             convertedPort.ContainerPort = internalPort.containerPort;
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp:672]: Defaulting to:
convertedPort.Family = AF_UNSPEC;
convertedPort.BindingAddress = nullptr;

If not specified by the caller. Runtime will need to interpretive these value appropriately,

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 5]
```diff
@@ -555,12 +557,42 @@ try
...
+            if (internalPort.windowsAddress != nullptr)
+            {
+                char addrBuf[INET6_ADDRSTRLEN]{};
+                switch (internalPort.windowsAddress->ss_family)
+                {
+                case AF_INET:
+                {
+                    const auto* addr4 = reinterpret_cast<const sockaddr_in*>(internalPort.windowsAddress);
+
+                    THROW_HR_IF_NULL(E_UNEXPECTED, inet_ntop(AF_INET, &addr4->sin_addr, addrBuf, sizeof(addrBuf)));
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: We need to provide more exacting errors via ETW.

**Review by JohnMcPMS — COMMENTED** [review]
Almost certainly needs a test update to prevent a negative test from failing when it passes in an address.  Not sure if we can expect IPv6 everywhere, but at least passing in `127.0.0.1` (or maybe IPv6 loopback works even if no external IPv6 address is available?).

[Edit: Search suggests that IPv6 loopback is always available if device supports IPv6.]

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 6]
```diff
@@ -671,12 +673,48 @@ try
...
+
+                    THROW_HR_IF_NULL(E_UNEXPECTED, inet_ntop(AF_INET6, &addr6->sin6_addr, addrBuf, sizeof(addrBuf)));
+
+                    convertedPort.Family = AF_INET6;
+                    break;
+                }
+
+                default:
+                    // Reject unsupported or malformed address families
+                    THROW_HR(E_INVALIDARG);
```
Comment (OneBlue) [inline src/windows/WslcSDK/wslcsdk.cpp]: nit: I recommend THROW_HR_MSG() so we can log the address family we received to make debugging easier

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 7]
```diff
@@ -671,12 +673,48 @@ try
...
+                }
+                bindingAddressStrings[i] = addrBuf;
+                convertedPort.BindingAddress = bindingAddressStrings[i].c_str();
+            }
+            else
+            {
+                // If no binding address is provided, default to wildcard.
+                convertedPort.Family = AF_UNSPEC;
```
Comment (OneBlue) [inline src/windows/WslcSDK/wslcsdk.cpp]: Won't this fail later if we leave AF_UNSPEC here ? 

I would recommend either defaulting to localhost, or requiring the caller to pass in the binding address. 

The later would be my preference, so that way there's no ambiguity

**Review by OneBlue — CHANGES_REQUESTED** [review]
Added a couple comments, I would also recommend: 

- Updating the pull request title
- Adding test coverage for the binding address wiring (at least for localhost, since that's the only thing that the service supports today)


You'll probably have to update existing tests as well

**Comment by benhillis** [comment]
Hey @1wizkid 👋 — Following up on this PR. Currently the **wslc tests are failing** in CI, and there are **11 unresolved review threads** with feedback from maintainers, including:
- Missing test coverage
- Dangling pointer concerns (flagged 4 times)
- Wildcard port binding security concern
- \\E_UNEXPECTED\\ error code usage
- \\AF_UNSPEC\\ failure risk

Is this change still needed? The review feedback covers some important safety and design issues that should be addressed before this can move forward. Let us know if you need any help!

**Comment by 1wizkid** [comment]
Ill focus on this today.

Thanks,
Richard

Sent from my T-Mobile 5G Device
Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 6 above)
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Added that code change to report family name

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 8]
```diff
@@ -623,9 +623,11 @@ try
     }
 
     std::unique_ptr<WSLCPortMapping[]> convertedPorts;
+    std::vector<std::string> bindingAddressStrings;
     if (internalContainerSettings->ports && internalContainerSettings->portsCount)
     {
         convertedPorts = std::make_unique<WSLCPortMapping[]>(internalContainerSettings->portsCount);
+        bindingAddressStrings.resize(internalContainerSettings->portsCount);
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Lifetime requirement noted in code.

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 9]
```diff
@@ -634,12 +636,48 @@ try
...
+
+                    convertedPort.Family = AF_INET;
+                    break;
+                }
+
+                case AF_INET6:
+                {
+                    const auto* addr6 = reinterpret_cast<const sockaddr_in6*>(internalPort.windowsAddress);
+
+                    THROW_HR_IF_NULL(E_UNEXPECTED, inet_ntop(AF_INET6, &addr6->sin6_addr, addrBuf, sizeof(addrBuf)));
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: code now properly converts result to corresponding hresult

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 10]
```diff
@@ -634,12 +636,48 @@ try
...
+                    convertedPort.Family = AF_INET6;
+                    break;
+                }
+
+                default:
+                    // Reject unsupported or malformed address families
+                    THROW_HR(E_INVALIDARG);
+                }
+                bindingAddressStrings[i] = addrBuf;
+                convertedPort.BindingAddress = bindingAddressStrings[i].c_str();
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Comment made in code to keep the two in scope

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 11]
```diff
@@ -634,12 +636,48 @@ try
...
+            }
+            else
+            {
+                // If no binding address is provided, default to wildcard.
+                convertedPort.Family = AF_UNSPEC;
+                convertedPort.BindingAddress = nullptr;
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Made this default to loopback.

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp:667]: Made comment in code to keep these two in same scope.

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp:672]: Code now defaults to local host

**Comment by 1wizkid** [comment]
The bulk of the issues addressed and new tests submitted as well that cover the usage of windowsAddress IPv4 & IPv6.

Waiting for the test pass to complete and will see if anything left to do.

Thanks,
Richard

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 12]
```diff
@@ -555,12 +557,32 @@ try
...
+                    convertedPort.Family = AF_INET6;
+                }
+                bindingAddressStrings[i] = addrBuf;
+                convertedPort.BindingAddress = bindingAddressStrings[i].c_str();
+            }
+            else
+            {
+                convertedPort.Family = AF_INET;
+                convertedPort.BindingAddress = "127.0.0.1";
+            }
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp:732]: tests added

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 7 above)
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: I went back and forth on this, but since family has to have something in it, I can't just set to zero which is AF_UNSPEC, I'm going to go ahead and default to AF_INET local host.

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 13]
```diff
@@ -671,14 +691,58 @@ try
...
+                    {
+                        THROW_HR_MSG(hr, "inet_ntop() failed for address family AF_INET6");
+                    }
+                    convertedPort.Family = AF_INET6;
+                    break;
+                }
+
+                default:
+                    // Reject unsupported or malformed address families
+                    THROW_HR(E_INVALIDARG);
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: change made

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 14]
```diff
@@ -671,14 +691,58 @@ try
...
+                }
+                HRESULT hr = strncpy_s(convertedPort.BindingAddress, sizeof(convertedPort.BindingAddress), addrBuf, _TRUNCATE);
+
+                if (hr == STRUNCATE)
+                {
+                    // Log this as a warning since the address is valid but was truncated to fit in the buffer, which means the port mapping may not work as expected.
+                }
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Implemented this change

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 15]
```diff
@@ -201,6 +201,25 @@ void EnsureAbsolutePath(const std::filesystem::path& path, bool containerPath)
...
+    }
+
+    switch (errno)
+    {
+    case EAFNOSUPPORT:
+        return E_INVALIDARG;
+    case ENOSPC:
+        return HRESULT_FROM_WIN32(ERROR_INSUFFICIENT_BUFFER);
+    default:
+        return E_FAIL;
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Implemented this change

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 7 above)
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: I've updated the title and added test coverage as well.

**Review by 1wizkid — COMMENTED** [review]
Feedback incorporated

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 16]
```diff
@@ -201,6 +201,26 @@ void EnsureAbsolutePath(const std::filesystem::path& path, bool containerPath)
     }
 }
 
+HRESULT InetNtopToHresult(int af, const void* src, char* dst, size_t dstCount)
+{
+    const char* result = inet_ntop(af, src, dst, dstCount);
+    if (result)
```
Comment (OneBlue) [inline src/windows/WslcSDK/wslcsdk.cpp]: I recommend throwing the win32 error directly: 

```
THROW_WIN32_IF(WSAGetLastError(), inet_ntop(af, src, dst, dstCount) == nullptr);
```

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 17]
```diff
@@ -671,14 +692,65 @@ try
...
+                    {
+                        THROW_HR_MSG(hr, "inet_ntop() failed for address family AF_INET6");
+                    }
+                    convertedPort.Family = AF_INET6;
+                    break;
+                }
+
+                default:
+                    // Reject unsupported or malformed address families
+                    THROW_HR(E_INVALIDARG);
```
Comment (OneBlue) [inline src/windows/WslcSDK/wslcsdk.cpp]: nit: I recommend logging the address family to make debugging easier

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 18]
```diff
@@ -671,14 +692,65 @@ try
...
-            // TODO: Ipv6 & custom binding address support.
-            convertedPort.Family = AF_INET;
-
             // TODO: Consider using standard protocol numbers instead of our own enum.
             convertedPort.Protocol = internalPort.protocol == WSLC_PORT_PROTOCOL_TCP ? IPPROTO_TCP : IPPROTO_UDP;
-            strcpy_s(convertedPort.BindingAddress, "127.0.0.1");
+            // Validate IP address if provided and if valid, copy to runtime structure.
+            if (internalPort.windowsAddress != nullptr)
+            {
+                char addrBuf[INET6_ADDRSTRLEN]{};
```
Comment (OneBlue) [inline src/windows/WslcSDK/wslcsdk.cpp]: Instead of having a temporary buffer, I recommend copying directly to `convertedPort.BindingAddress`

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 19]
```diff
@@ -809,7 +881,11 @@ try
 
     for (uint32_t i = 0; i < portMappingCount; ++i)
     {
-        RETURN_HR_IF(E_NOTIMPL, portMappings[i].windowsAddress != nullptr);
+        if (portMappings[i].windowsAddress != nullptr)
+        {
+            const auto family = portMappings[i].windowsAddress->ss_family;
+            RETURN_HR_IF(E_INVALIDARG, family != AF_INET && family != AF_INET6);
```
Comment (OneBlue) [inline src/windows/WslcSDK/wslcsdk.cpp]: nit: I recommend logging the invalid address family here

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 16 above)
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Making this change. Leaving in it own function to simplify updates if needed.

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 18 above)
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Change made

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 17 above)
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Change made.

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 19 above)
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Change made

File: src/windows/WslcSDK/wslcsdk.cpp (hunk 16 above)
Comment (OneBlue) [reply src/windows/WslcSDK/wslcsdk.cpp]: Was the change pushed ? I still see the old diff on this PR

File: test/windows/WslcSdkTests.cpp [hunk 20]
```diff
@@ -844,18 +844,18 @@ class WslcSdkTests
 
         // Negative: port mappings with NONE networking must fail at container creation.
         {
-            WslcContainerSettings containerSettings;
-            VERIFY_SUCCEEDED(WslcInitContainerSettings("debian:latest", &containerSettings));
-            VERIFY_SUCCEEDED(WslcSetContainerSettingsNetworkingMode(&containerSettings, WSLC_CONTAINER_NETWORKING_MODE_NONE));
+            WslcContainerSettings containerSettings1;
```
Comment (OneBlue) [inline test/windows/WslcSdkTests.cpp:778]: nit: this doesn't need to be renamed since it's in a different scope (same for below)

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 21]
```diff
@@ -201,6 +201,12 @@ void EnsureAbsolutePath(const std::filesystem::path& path, bool containerPath)
     }
 }
 
+HRESULT InetNtopToHresult(int af, const void* src, char* dst, size_t dstCount)
+{
+    THROW_WIN32_IF(static_cast<unsigned long>(WSAGetLastError()), inet_ntop(af, src, dst, dstCount) == nullptr);
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Yes, I will return the HRESULT from this function so caller can also issue proper error logging

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 22]
```diff
@@ -671,14 +678,69 @@ try
             convertedPort.HostPort = internalPort.windowsPort;
             convertedPort.ContainerPort = internalPort.containerPort;
 
-            // TODO: Ipv6 & custom binding address support.
-            convertedPort.Family = AF_INET;
-
             // TODO: Consider using standard protocol numbers instead of our own enum.
             convertedPort.Protocol = internalPort.protocol == WSLC_PORT_PROTOCOL_TCP ? IPPROTO_TCP : IPPROTO_UDP;
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: change made

[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Change made

File: test/windows/WslcSdkTests.cpp [hunk 23]
```diff
@@ -894,6 +894,105 @@ class WslcSdkTests
...
+            sockaddr_storage badAddr{};
+            badAddr.ss_family = AF_UNIX; // unsupported for port mapping
+
+            WslcContainerPortMapping mapping{};
+            mapping.windowsPort = 12345;
+            mapping.containerPort = 8000;
+            mapping.protocol = WSLC_PORT_PROTOCOL_TCP;
+            mapping.windowsAddress = &badAddr;
+            VERIFY_ARE_EQUAL(WslcSetContainerSettingsPortMappings(&containerSettings5, &mapping, 1), E_INVALIDARG);
+        }
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply test/windows/WslcSdkTests.cpp]: change to comment made

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 24]
```diff
@@ -671,14 +680,79 @@ try
...
+                {
+                    const auto* addr6 = reinterpret_cast<const sockaddr_in6*>(internalPort.windowsAddress);
+                    HRESULT hr = InetNtopToHresult(AF_INET6, &addr6->sin6_addr, convertedPort.BindingAddress, sizeof(convertedPort.BindingAddress));
+                    if (FAILED(hr))
+                    {
+                        const auto* b = addr6->sin6_addr.u.Byte;
+                        THROW_HR_MSG(
+                            hr,
+                            "inet_ntop() failed for AF_INET6 address: % 02x % 02x : % 02x % 02x : % 02x % 02x : % 02x % 02x : % "
+                            "02x % 02x : % 02x % 02x : % 02x % 02x : % 02x %02x ",
```
Comment (OneBlue) [inline src/windows/WslcSDK/wslcsdk.cpp]: this formatting string doesn't actually work because of the spaces between the "%" and the specifier. 

I'd recommend just dropping the logging of the IP buffer since all buffers are actually valid ip addresses, realistically the only way `inet_ntop` could fail would be if we passed an invalid address family or buffer size

Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: Will drop it. CLANG formatting keeps messing up what I type in...

File: test/windows/WslcSdkTests.cpp (hunk 20 above)
Comment (1wizkid) [reply test/windows/WslcSdkTests.cpp:778]: I did this "numbering" of the containerSettings var because in the testing output I couldn't tell what was failing because the line it pointed it is identical. Doing this allows me to find the exact failing function because the var name it displays has the # appended to it.

Comment (OneBlue) [reply test/windows/WslcSdkTests.cpp:778]: TE outputs the line number of the failure, that's how we usually resolve situations like these

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 25]
```diff
@@ -809,10 +862,15 @@ try
...
-        RETURN_HR_IF(E_NOTIMPL, portMappings[i].windowsAddress != nullptr);
-        RETURN_HR_IF(E_NOTIMPL, portMappings[i].protocol != 0);
+        if (portMappings[i].windowsAddress != nullptr)
+        {
+            const auto family = portMappings[i].windowsAddress->ss_family;
+            RETURN_HR_IF_MSG(
+                E_INVALIDARG, family != AF_INET && family != AF_INET6, "Unsupported address family: %d at port mapping index %u", family, i);
+        }
+        RETURN_HR_IF_MSG(
+            E_NOTIMPL, portMappings[i].protocol != 0, "Unsupported protocol: %d at port mapping index %u", portMappings[i].protocol, i);
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp:871]: We don't support UDP at the moment and will enforce this at the Set*() point. Will leave as is.

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 26]
```diff
@@ -671,14 +680,58 @@ try
...
+                    break;
+                }
+
+                case AF_INET6:
+                {
+                    const auto* addr6 = reinterpret_cast<const sockaddr_in6*>(internalPort.windowsAddress);
+                    HRESULT hr = InetNtopToHresult(AF_INET6, &addr6->sin6_addr, convertedPort.BindingAddress, sizeof(convertedPort.BindingAddress));
+                    if (FAILED(hr))
+                    {
+                        THROW_HR_MSG(hr,"inet_ntop() failed for AF_INET6 address");
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: fixed

File: src/windows/WslcSDK/wslcsdk.cpp [hunk 27]
```diff
@@ -200,6 +200,14 @@ void EnsureAbsolutePath(const std::filesystem::path& path, bool containerPath)
...
     }
 }
+HRESULT InetNtopToHresult(int af, const void* src, char* dst, size_t dstCount)
+{
+    if (inet_ntop(af, src, dst, dstCount) == nullptr)
+    {
+        return HRESULT_FROM_WIN32(WSAGetLastError());
+    }
+    return S_OK;
+}
```
[the remark this answers is not in the document]
Comment (1wizkid) [reply src/windows/WslcSDK/wslcsdk.cpp]: This is an internal only function. will mark as static.

**Review by OneBlue — APPROVED** [review]

File: diagnostics/trace01.etl
Comment (JohnMcPMS) [inline diagnostics/trace01.etl:1]: Remove .etl files.

Comment (OneBlue) [reply diagnostics/trace01.etl:1]: Good catch ! Not sure how I missed this

**Review by OneBlue — CHANGES_REQUESTED** [review]
Let's remove the .etl file before merging

**Review by JohnMcPMS — APPROVED** [review]

**Review by OneBlue — APPROVED** [review]
